### PR TITLE
Add a matching pattern

### PR DESCRIPTION
**Pull request by Boris-2021**
In the configuration file drain3.ini. Add configuration parameters.
`matched_pattern = JaccardDrain`

"JaccardDrain" means matching text with Jaccard coefficients.

### CODE REVIEW

**Comment by Boris-2021** [comment]
In the configuration file drain3.ini. Add configuration parameters.
`matched_pattern = JaccardDrain`
"JaccardDrain" means matching text with Jaccard coefficients.

**Comment by davidohana** [comment]
Hi, thank you for the PR!
Can you please explain the background and motivation for this PR?

**Comment by Boris-2021** [comment]
> Hi, thank you for the PR! Can you please explain the background and motivation for this PR?


Hi, glad you reviewed my PR.



### background :

I used Drian3 project to apply in a microservice background, using log parsing  to raise alarms for background stability.

In the process of using Drain3, I realized that its algorithm is essentially a process of clustering log messages by text similarity.In this process, the use of prior knowledge to propose variable factors into 

```
<IP>;<NUM>;<*>...
```

can improve the data and make it better clustered. The design of tree structure storage of log text can ensure the accuracy of classification and improve the retrieval speed.

In the current clustering process, the log length is used to classify the log at the first step, and the text similarity is calculated in the final node to further divide the cluster. However, in the process of use, because of the variety of log variables, it is possible to change the length of the log, so that the log describing the same event is classified into different classes according to the length.

（A convenient solution is to enhance the regularization rule so that the algorithm can better match variables of different lengths to the same length. This requires the user to do specific data analysis on the data to design specific regular expressions, but I think there should be a smarter way.）



### motivation：

Based on the above problem, I created a new class JaccardDrain, which inherits from Drain. Rewrite some methods in Drain. It changes the process of text clustering. Mainly based on the following ideas:

**1-Node partition:**

The first node of the tree is no longer divided by the length of the log message, but the first word is the first node (and subsequent nodes are based on the word at the next position). The reason for this is based on the assumption that the first few letters of most log messages are fixed.

**2-New similarity calculation method:**

The original formula for calculating the similarity between log messages and templates requires that both sides have the same length, and then compares the proportion of the same number of characters in the length. 

The new calculation method can calculate the proportion of the number of strings with the same log information and template in the total number of strings without requiring the same length( A∩B / A∪B ). This formula gives a higher score when two texts are closer in length and content. There is no restriction that the computed text must be of the same length.

**3-how to update the template:**

When new log information is located in a log template cluster. When the original template is updated, there are two cases. One is to check the position of the new variable from the new log information and update it to

```
<*>
```

The other is to check that the length of the word in the variable becomes longer in the new log message and update it to

```
<*> <*> ...
```


Based on these improvements, the log template can be clustered more intelligently (no need to consider the influence of the length of the template variable, no need to set the regular expression about the length of the variable in advance).

**Comment by Boris-2021** [comment]
Hi, glad you reviewed my PR.

background :

I used Drian3 project to apply in a microservice background, using log parsing to raise alarms for background stability.

In the process of using Drain3, I realized that its algorithm is essentially a process of clustering log messages by text similarity.In this process, the use of prior knowledge to propose variable factors into
<IP&gt;;<NUM&gt;;<*&gt;... 

can improve the data and make it better clustered. The design of tree structure storage of log text can ensure the accuracy of classification and improve the retrieval speed.

In the current clustering process, the log length is used to classify the log at the first step, and the text similarity is calculated in the final node to further divide the cluster. However, in the process of use, because of the variety of log variables, it is possible to change the length of the log, so that the log describing the same event is classified into different classes according to the length.

（A convenient solution is to enhance the regularization rule so that the algorithm can better match variables of different lengths to the same length. This requires the user to do specific data analysis on the data to design specific regular expressions, but I think there should be a smarter way.）

motivation：

Based on the above problem, I created a new class JaccardDrain, which inherits from Drain. Rewrite some methods in Drain. It changes the process of text clustering. Mainly based on the following ideas:

1-Node partition:

The first node of the tree is no longer divided by the length of the log message, but the first word is the first node (and subsequent nodes are based on the word at the next position). The reason for this is based on the assumption that the first few letters of most log messages are fixed.

2-New similarity calculation method:

The original formula for calculating the similarity between log messages and templates requires that both sides have the same length, and then compares the proportion of the same number of characters in the length.

The new calculation method can calculate the proportion of the number of strings with the same log information and template in the total number of strings without requiring the same length( A∩B / A∪B ). This formula gives a higher score when two texts are closer in length and content. There is no restriction that the computed text must be of the same length.

3-how to update the template:

When new log information is located in a log template cluster. When the original template is updated, there are two cases. One is to check the position of the new variable from the new log information and update it to
<*&gt; 

The other is to check that the length of the word in the variable becomes longer in the new log message and update it to
<*&gt; <*&gt; ... 

Based on these improvements, the log template can be clustered more intelligently (no need to consider the influence of the length of the template variable, no need to set the regular expression about the length of the variable in advance).



Boris


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "IBM/Drain3"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年8月11日(星期四) 晚上9:26
***@***.***&gt;;
抄送:&nbsp;"无产者 ***@***.******@***.***&gt;;
主题:&nbsp;Re: [IBM/Drain3] Add a matching pattern (PR #67)





 
Hi, thank you for the PR!
 Can you please explain the background and motivation for this PR?

File: drain3/template_miner.py [hunk 1]
```diff
@@ -44,13 +44,21 @@ def __init__(self,
...
         param_str = self.config.mask_prefix + "*" + self.config.mask_suffix
-        self.drain = Drain(
+
+        # Follow the configuration in the configuration file to instantiate Drain
+        target_obj = self.config.matched_pattern
+        if target_obj not in ["Drain","JaccardDrain"]:
+            raise ValueError(f"Invalid matched_pattern: {target_obj}, must be either 'Drain' or 'JaccardDrain'")
+
+        '''
+        self.drain = JaccardDrain(
```
Comment (davidohana) [inline drain3/template_miner.py:61]: Why did you make JaccardDrain the default? Default  behavior should stay exactly as it is. JaccardDrain should be opt-in.

File: examples/drain3.ini [hunk 2]
```diff
@@ -16,8 +16,9 @@ mask_prefix = <:
 mask_suffix = :>
 
 [DRAIN]
+matched_pattern = JaccardDrain
 sim_th = 0.4
-depth = 4
+depth = 3
```
Comment (davidohana) [inline examples/drain3.ini:21]: Again, we should not change the default behavior unless there is a good reason.

File: examples/drain3.ini [hunk 3]
```diff
@@ -16,8 +16,9 @@ mask_prefix = <:
 mask_suffix = :>
 
 [DRAIN]
+matched_pattern = JaccardDrain
```
Comment (davidohana) [inline examples/drain3.ini:19]: Not sure `matched_pattern` is a good name here. Perhaps call it `engine`?

File: drain3/template_miner.py (hunk 1 above)
Comment (Boris-2021) [reply drain3/template_miner.py:61]: You are right, Drain should be the default, this is the point I missed in code.

File: examples/drain3.ini (hunk 2 above)
Comment (Boris-2021) [reply examples/drain3.ini:21]: ok, default behavior same as before is better.

Comment (Boris-2021) [reply examples/drain3.ini:21]: ok, default behavior same as before is better.

File: examples/drain3.ini (hunk 3 above)
Comment (Boris-2021) [reply examples/drain3.ini:19]: That sounds good, I agree.

**Comment by davidohana** [comment]
Thank you. We will require that unit tests will pass also with the Jaccard Engine. Also a sample with a real-world log  (similar to the big file demo) in which Jaccard is better than the original Drain. Preferably with comparison to Drain.

File: drain3/drain.py [hunk 4]
```diff
@@ -419,3 +419,231 @@ def get_total_cluster_size(self):
         for c in self.id_to_cluster.values():
             size += c.size
         return size
+
+
+class JaccardDrain(Drain):
```
Comment (davidohana) [inline drain3/drain.py:424]: Please split this class into another file.

**Comment by davidohana** [comment]
Also, I am concerned about the amount of repeated code in JaccardDrain. Can you try to inherit more code and functions from the Drain class?

**Comment by Superskyyy** [comment]
@Boris-2021 Hello, a gentle reminder on the progress of this PR. I appreciate it if you have time to finish this and merge it into the main branch.

File: drain3/drain.py (hunk 4 above)
Comment (Boris-2021) [reply drain3/drain.py]: Okay, that's done

**Comment by Boris-2021** [comment]
> Also, I am concerned about the amount of repeated code in JaccardDrain. Can you try to inherit more code and functions from the Drain class?

Ok, I've optimized some code so that JaccardDrain inherits as much code as possible, and now it only needs to rewrite 5 functions.

**Comment by Boris-2021** [comment]
> @Boris-2021 Hello, a gentle reminder on the progress of this PR. I appreciate it if you have time to finish this and merge it into the main branch.


Ok, I'm sorry for the delay, it's done now

**Comment by Superskyyy** [comment]
Thank you! I will take a look over this weekend.

**Comment by Impelon** [comment]
If I understand correctly, `extract_parameters()` will not work correctly with the templates produced by JaccardDrain, because the algorithm assumes that template and the line to match to have the same amount of tokens.
However, from the examples it becomes clear that
```
Received disconnect from 127.0.0.1: 1234: [preauth]
```
could be matched against
```
Received disconnect from <:IP:>: <:NUM:>: <:*:> <:*:> <:*:> <:*:> <:*:> <:*:> <:*:>
```
(which obviously does not have the same amount of tokens.) Please correct me, if I misunderstood! :)

<details>
<summary>In case I understood everything correctly so far, open this spoiler to read on. Otherwise ignore the rest of this comment!</summary>

Currently, the extraction algorithm would notice that the template and message do not match up and return `None`:
```python3
>>> repr(miner.extract_parameters("Received disconnect from <:IP:>: <:NUM:>: <:*:> <:*:> <:*:> <:*:> <:*:> <:*:> <:*:>", "Received disconnect from 127.0.0.1: 1234: [preauth]"))
'None'
>>> repr(miner.extract_parameters("Received disconnect from <:IP:>: <:NUM:>: <:*:>", "Received disconnect from 127.0.0.1: 1234: [preauth]"))
"[ExtractedParameter(value='127.0.0.1', mask_name='IP'), ExtractedParameter(value='1234', mask_name='NUM'), ExtractedParameter(value='[preauth]', mask_name='*')]"
```

The problem is that a wildcard pattern can now be left without a match, where as before each wildcard necessarily matched up with a parameter.
I am not sure how to best solve this, but I do have some options in mind. Probably they should be discussed in another PR though, so the discussion in this PR can focus on the new implementation.

I'm not suggesting to delay the merging of this PR, but intended to raise awareness to the resulting problem to parameter extraction. I'd say it's not a major issue to not have the extraction fully working, because this will only happen to users who explicitly wish to switch to JaccardDrain.
Maybe it would be nice to note somewhere that parameter extraction does not work for the most part with JaccardDrain until a fix is provided.
</details>

**Comment by Superskyyy** [comment]
Generally LGTM, let's fix the mentioned problem in the next PR.

**Review by Superskyyy — APPROVED** [review]

**Pull request merged**